### PR TITLE
Single Distro API

### LINKED ISSUE #10695: Address issues of the POST /distributions API endpoint
<strike>The _POST /distributions API endpoint_ doesn't work.</strike> _It was fixed in #10762._

The _POST /distributions API endpoint_ is not documented in the old and new API documentations.

The _PUT /distributions_ API endpoint is already handling creation of distributions. This is confusing as _PUT_ endpoints are normally used to update data.

We need to refactor both endpoints to have the _POST_ only create distributions and the _PUT_ only update distributions. This would need to be done in a new API version.

### PR DESCRIPTION
**Pull request by hennevogel**
Makes it possible to CRUD a single distribution.

Fixes #10695

### CODE REVIEW

**Comment by hennevogel** [comment]
Dependent on #11050

File: src/api/app/models/distribution.rb [hunk 1]
```diff
@@ -4,25 +4,35 @@ class Distribution < ApplicationRecord
...
+
+    architectures.clear
+    xmlhash.elements('architecture') do |architecture_name|
+      architecture = Architecture.find_by(name: architecture_name)
+      architectures << architecture if architecture
+    end
+
+    icons.clear
+    xmlhash.elements('icon') do |icon|
+      icons.create(width: icon['width'], height: icon['height'], url: icon['url'])
```
Comment (saraycp) [inline src/api/app/models/distribution.rb:32]: Here you could also use the hash `icon`.

**Comment by hennevogel** [comment]
Here is how you can try this:

Create some distributions on your rails console:
```
[1] pry(main)> include FactoryBot::Syntax::Methods
=> Object
[2] pry(main)> create_list(:distribution, 5)
....
```

With osc against your local setup:

Index: `osc api /distributions`
Show: `osc api /distributions/5`
Update: `osc api -e /distributions/5`
Create:
```
$ osc api /distributions/5 > distro.xml
$ vi distro.xml
... change some attributes...
$ osc api -X POST -T distro.xml /distributions
```
Bulk Update: `osc api -e /distributions/bulk_replace`

**Comment by adrianschroeter** [comment]
On Donnerstag, 20. Mai 2021, 16:56:08 CEST Henne Vogelsang wrote:
> Here is how you can try this:
> 
> Create some distributions on your rails console:
> ```
> [1] pry(main)> include FactoryBot::Syntax::Methods
> => Object
> [2] pry(main)> create_list(:distribution, 5)
> ....

not used by admins as far as I am aware off

> ```
> 
> With osc against your local setup:
> 
> Index: `osc api /distributions`
> Show: `osc api /distributions/5`
> Update: `osc api -e /distributions/5`

also not used by admins afaik.

> Create:
> ```
> $ osc api /distributions/5 > distro.xml
> $ vi distro.xml
> ... change some attributes...
> $ osc api -X POST -T distro.xml /distributions
> ```
> Bulk Update: `osc api -e /distributions/bulk_replace`

Admins use always for this

 # osc api -e /distributions

which is a PUT, can't we stay compatible here? Is it really only 
for internal needs that we break the API here?

Couldn't you introduce a new route for the new behaviour instead in addition?


-- 

Adrian Schroeter ***@***.***>
Build Infrastructure Project Manager

SUSE Software Solutions Germany GmbH,  Maxfeldstr. 5, 90409 Nuernberg, Germany
(HRB 247165, AG München), Geschäftsführer: Felix Imendörffer

**Comment by adrianschroeter** [comment]
On Donnerstag, 20. Mai 2021, 17:11:47 CEST Adrian Schröter wrote:
> On Donnerstag, 20. Mai 2021, 16:56:08 CEST Henne Vogelsang wrote:
> > Here is how you can try this:
> > 
> > Create some distributions on your rails console:
> > ```
> > [1] pry(main)> include FactoryBot::Syntax::Methods
> > => Object
> > [2] pry(main)> create_list(:distribution, 5)
> > ....
> 
> not used by admins as far as I am aware off
> 
> > ```
> > 
> > With osc against your local setup:
> > 
> > Index: `osc api /distributions`
> > Show: `osc api /distributions/5`
> > Update: `osc api -e /distributions/5`
> 
> also not used by admins afaik.
> 
> > Create:
> > ```
> > $ osc api /distributions/5 > distro.xml
> > $ vi distro.xml
> > ... change some attributes...
> > $ osc api -X POST -T distro.xml /distributions
> > ```
> > Bulk Update: `osc api -e /distributions/bulk_replace`
> 
> Admins use always for this
> 
>  # osc api -e /distributions
> 
> which is a PUT, can't we stay compatible here? Is it really only 
> for internal needs that we break the API here?
> 
> Couldn't you introduce a new route for the new behaviour instead in addition?

also our official admin documentation uses the PUT route to /distributions btw.

-- 

Adrian Schroeter ***@***.***>
Build Infrastructure Project Manager

SUSE Software Solutions Germany GmbH,  Maxfeldstr. 5, 90409 Nuernberg, Germany
(HRB 247165, AG München), Geschäftsführer: Felix Imendörffer

**Comment by hennevogel** [comment]
>>  Update: `osc api -e /distributions/5`

> also not used by admins afaik.

They couldn't if they would want to. It doesn't work.

> Admins use always for this
>
> osc api -e /distributions

And to update the list of distribution you can use `osc api -e /distributions/bulk_replace` in the same manner now.

> Is it really only for internal needs that we break the API here?

"only" yes. This simplifies the code and the routes very much and hence makes it understandable for our future selves. The cost is that admins now have to append `/bulk_replace` to their command line interface. Seems reasonable to me.

> Couldn't you introduce a new route for the new behaviour instead in addition?

No because we are leaving the rails way but using `PUT /distributions` for bulk updates instead of resource creation. And once the :steam_locomotive:  leaves the rails...

File: src/api/spec/controllers/distributions_controller_spec.rb [hunk 2]
```diff
@@ -32,30 +53,90 @@
...
+      subject { post :create, body: distributions_xml, format: :xml }
 
       it { expect { subject }.not_to change(Distribution, :count) }
       it { is_expected.to have_http_status(:bad_request) }
     end
+  end
 
-    context 'when xml is empty' do
-      subject { post :create, format: :xml }
+  describe '#update' do
```
[the remark this answers is not in the document]
Comment (hennevogel) [reply src/api/spec/controllers/distributions_controller_spec.rb:74]: What is it you want to test? There is no logic...

[the remark this answers is not in the document]
Comment (krauselukas) [reply src/api/spec/controllers/distributions_controller_spec.rb:74]: okay, yeah its just calling the model, fine

**Review by krauselukas — COMMENTED** [review]
Besides what @saraycp commented not much more to add regarding the code. I only realized that the `Distributions` endpoints are already documented in our new OpenAPI documentation, so those need to be updated as well, or at least we should track that they need to be updated

File: docs/api/api/api.txt [hunk 3]
```diff
@@ -83,38 +83,42 @@ XmlResult: issue
 
 GET /distributions
 
-  Get the list of base distributions hosted on this OBS instance
+  Get all base distributions hosted on this OBS instance
```
Comment (saraycp) [inline docs/api/api/api.txt:86]:
```suggestion
  Get all the base distributions hosted on this OBS instance
```

File: docs/api/api/api.txt [hunk 4]
```diff
@@ -83,38 +83,42 @@ XmlResult: issue
...
-
 GET /distributions/include_remotes
 
   Get the list of base distributions hosted on this OBS instance and on all used remote instances
 
 XmlResult: distributions
 
+POST /distributions/bulk_replace
+
+  Replace all base distributions hosted on this OBS instance
```
Comment (saraycp) [inline docs/api/api/api.txt:118]:
```suggestion
  Replace all the base distributions hosted on this OBS instance
```

**Review by saraycp — COMMENTED** [review]
After manually testing the endpoints, LGTM.
Just a few little improvements I leave in the comments.

**Comment by adrianschroeter** [comment]
On Freitag, 21. Mai 2021, 13:48:46 CEST Saray Cabrera Padrón wrote:
> @saraycp commented on this pull request.
> 
> After manually testing the endpoints, LGTM.
> Just a few little improvements I leave in the comments. 
> 
> > @@ -83,38 +83,42 @@ XmlResult: issue
>  
>  GET /distributions
>  
> -  Get the list of base distributions hosted on this OBS instance
> +  Get all base distributions hosted on this OBS instance
> 
> ```suggestion
>   Get all the base distributions hosted on this OBS instance
> ```
> 
> >  GET /distributions/include_remotes
>  
>    Get the list of base distributions hosted on this OBS instance and on all used remote instances
>  
>  XmlResult: distributions
>  
> +POST /distributions/bulk_replace
> +
> +  Replace all base distributions hosted on this OBS instance

please, do not invent such names no one can remember and incompatibilities
which are not needed.

Just keep PUT to /distributions working as last > 10 years, even
when it needs another line of extra code.

Please accept this from your project manager and also the main user of
this specific route ;)

-- 

Adrian Schroeter ***@***.***>
Build Infrastructure Project Manager

SUSE Software Solutions Germany GmbH,  Maxfeldstr. 5, 90409 Nuernberg, Germany
(HRB 247165, AG München), Geschäftsführer: Felix Imendörffer

**Comment by hennevogel** [comment]
> please, do not invent such names no one can remember

So the only thing you can remember is the thing that exists? C'mon...

> and incompatibilities which are not needed

But they **are** needed to make this code more maintainable a.k.a. easier to understand, faster to extend etc. 

> even when it needs another line of extra code

It doesn't need "another line of extra code". It needs many extra lines of code as it changes all the free/cheap things the framework provides to us to CRUD resources by putting bulk updates in the place of creation. You have to have custom routes, you have to have custom actions, you have to have developer documentation why this is so etc.

If we can't even invest this little for this "admin" route that really only has one user, *you*, how are we ever going to evolve the API beyond what is there? :thinking:

**Comment by hennevogel** [comment]
Addressed the feedback, brought back the compat route and documented with swagger. Please review again.

**Comment by hennevogel** [comment]
#11050 is merged

**Review by saraycp — APPROVED** [review]
For me, both the `bulk_replace` and the brought back the route work fine.

**Comment by dmarcoux** [comment]
So changes from this PR are related to #10695, right?

**Comment by hennevogel** [comment]
here goes nothing

**Pull request merged**